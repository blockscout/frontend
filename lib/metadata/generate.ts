import type { ApiData, Metadata } from './types';

import type { Route } from 'nextjs-routes';

import config from 'configs/app';
import getNetworkTitle from 'lib/networks/getNetworkTitle';

import compileValue from './compileValue';
import getPageOgType from './getPageOgType';
import * as templates from './templates';

export default function generate<R extends Route>(route: R, apiData?: ApiData<R>): Metadata {
  const params = {
    ...route.query,
    ...apiData,
    network_name: config.chain.name,
    network_title: getNetworkTitle(),
  };

  const title = 'â–¼ Lux Testnet Blockchain Explorer';
  const description = compileValue(templates.description.make(route.pathname), params);

  const pageOgType = getPageOgType(route.pathname);

  return {
    title: title,
    description,
    opengraph: {
      title: title,
      description: pageOgType !== 'Regular page' ? config.meta.og.description : '',
      imageUrl: pageOgType !== 'Regular page' ? config.meta.og.imageUrl : '',
    },
  };
}
