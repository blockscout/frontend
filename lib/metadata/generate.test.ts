import type { ApiData } from './types';

import type { Route } from 'nextjs-routes';

import generate from './generate';

interface TestCase<Pathname extends Route['pathname']> {
  title: string;
  route: {
    pathname: Pathname;
    query?: Route['query'];
  };
  apiData?: ApiData<Pathname>;
}

const TEST_CASES = [
  {
    title: 'static route',
    route: {
      pathname: '/txs',
    },
  } as TestCase<'/txs'>,
  {
    title: 'dynamic route',
    route: {
      pathname: '/tx/[hash]',
      query: { hash: '0x12345' },
    },
  } as TestCase<'/tx/[hash]'>,
  {
    title: 'dynamic route with API data',
    route: {
      pathname: '/token/[hash]',
      query: { hash: '0x12345' },
    },
    apiData: { symbol: 'USDT' },
  } as TestCase<'/token/[hash]'>,
];

describe('generates correct metadata for:', () => {
  TEST_CASES.forEach((testCase) => {
    it(`${ testCase.title }`, () => {
      const result = generate(testCase.route, testCase.apiData);
      expect(result).toMatchSnapshot();
    });
  });
});
