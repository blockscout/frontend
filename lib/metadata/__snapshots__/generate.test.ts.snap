// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates correct metadata for: dynamic route 1`] = `
{
  "canonical": undefined,
  "description": "View transaction 0x12345 on Blockscout (Blockscout) Explorer",
  "opengraph": {
    "description": "",
    "imageUrl": "",
    "title": "Blockscout transaction 0x12345 | Blockscout",
  },
  "title": "Blockscout transaction 0x12345 | Blockscout",
}
`;

exports[`generates correct metadata for: dynamic route with API data 1`] = `
{
  "canonical": undefined,
  "description": "0x12345, balances and analytics on the Blockscout (Blockscout) Explorer",
  "opengraph": {
    "description": "",
    "imageUrl": "",
    "title": "Blockscout USDT token details | Blockscout",
  },
  "title": "Blockscout USDT token details | Blockscout",
}
`;

exports[`generates correct metadata for: static route 1`] = `
{
  "canonical": "http://localhost:3000/txs",
  "description": "Open-source block explorer by Blockscout. Search transactions, verify smart contracts, analyze addresses, and track network activity. Complete blockchain data and APIs for the Blockscout (Blockscout) Explorer network.",
  "opengraph": {
    "description": "",
    "imageUrl": "http://localhost:3000/static/og_placeholder.png",
    "title": "Blockscout transactions - Blockscout explorer | Blockscout",
  },
  "title": "Blockscout transactions - Blockscout explorer | Blockscout",
}
`;
